refactor(parser): remove lookahead for parsing import declarations and specifiers

camchenry opened the pull request:
- towards https://github.com/oxc-project/oxc/issues/11334

Rewrites the parsing for import declarations to not require any lookahead. I've diverged from the TS reference implementation here because it uses a fair amount of `tryParse` and lookahead. The main trick used here is simply parsing as we go and then doing "lookback" to see if the identifier we just parsed was actually a keyword like `defer` or `type`. This allows us to not need to use rewind at all.

Also improves diagnostics slightly by emitting a new diagnostic message for `import ... =`. In the end, this didn't improve conformance at all, but it's a perf win of about 1% on some files.

camchenry commented:
* **#11381** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11381?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11381?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11381).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry commented:
Benchmark shows +1% perf improvement in cal.com.tsx parsing, and negligible impact elsewhere.

Boshen commented:
> It looks like there is a slight regression in `tasks/coverage/snapshots/estree_typescript.snap`? Though I'm not really sure how to interpret what's wrong there. If anyone else understands it, please let me know.

Here's how I debug this:

* for the failed file tasks/coverage/typescript/tests/cases/conformance/externalModules/typeOnly/importDefaultNamedType.ts 
* run `just watch-example codegen tasks/coverage/typescript/tests/cases/conformance/externalModules/typeOnly/importDefaultNamedType.ts`
* See output `import "./a";` vs orignal (`import type from './a';`)

So this special case is missed :-)

camchenry commented:
@Boshen This should be all fixed up now. Still a 1% perf improvement for the cal.com.tsx parsing benchmark.

Boshen commented:
Nice work!

Boshen commented:
### Merge activity

* **Jun 2, 2:43 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camchenry committed:
refactor(parser): remove lookahead for parsing import declarations and specifiers (#11381)

- towards https://github.com/oxc-project/oxc/issues/11334

Rewrites the parsing for import declarations to not require any lookahead. I've diverged from the TS reference implementation here because it uses a fair amount of `tryParse` and lookahead. The main trick used here is simply parsing as we go and then doing "lookback" to see if the identifier we just parsed was actually a keyword like `defer` or `type`. This allows us to not need to use rewind at all.

Also improves diagnostics slightly by emitting a new diagnostic message for `import ... =`. In the end, this didn't improve conformance at all, but it's a perf win of about 1% on some files.
